promql/engine.go

```diff
@@ -924,6 +932,7 @@ func (ev *evaluator) rangeEval(f func([]parser.Value, *EvalNodeHelper) (Vector,
 		enh.Out = result[:0] // Reuse result vector.
 		warnings = append(warnings, ws...)
 
+		remainingSamples := ev.maxSamples - ev.currentSamples
```

Comment:
It'd be better to inline this, as you do below.

Reply:
I see two ways. Do you see something else or have a preference?

```
min(numSteps, ev.maxSamples - ev.currentSamples + len(result))
```

or (re)move the `ev.currentSamples += len(result)` to a later point and deal with everything that this implies, such as changing the `if current + len(result) > maxSamples { abort... }`

Reply:
Why can't it be ev.maxSamples - ev.currentSamples?

Reply:
Because of the line immediately following `remainingSamples` where we are adding len(result) to `ev.currentSamples`? I assume this is done to check if max samples is reached only once instead of on every loop iteration and before the other fast path.


Is my reading correct? If we start with:

  * ev.maxSamples == 10
  * ev.currentSamples = 0
 *  len(result) == 10.

 -> ev.currentSamples += len(result) "10"   == 10
 now in the loop (let's assume a single metric)
 -> Points: getPointSlice(min(numSteps, 10 - 10)
 -> Then ss.Points = append(ss.Points, point) will grow the slice.


Did I miss anything?

Reply:
If we're right up to the edge of the limit I wouldn't worry too much about getting it perfect, we're probably about to hit the limit anyway.

This would however result in far more capacity being allocated than you'd want if there's a high cardinality result as each of them would get the same value rather than their fraction of it.